### PR TITLE
Use MQTT_NS::error_code instead of boost::system::error_code for error_codes

### PR DESCRIPTION
**Pull request by jonesmz**
Replaces https://github.com/redboltz/mqtt_cpp/pull/525

### CODE REVIEW

File: include/mqtt/client.hpp [hunk 1]
```diff
@@ -1054,7 +1054,7 @@ class client : public endpoint<std::mutex, std::lock_guard, PacketIdBytes> {
         any session_life_keeper,
         boost::system::error_code& ec) {
         start_session(force_move(props), force_move(session_life_keeper));
-        ec = boost::system::errc::make_error_code(boost::system::errc::success);
+        ec = {};
```
Comment (redboltz) [inline include/mqtt/client.hpp]: It is difficult to understand. Could you revert it as the original code?

File: include/mqtt/endpoint.hpp [hunk 2]
```diff
@@ -6760,18 +6761,16 @@ class endpoint : public std::enable_shared_from_this<endpoint<Mutex, LockGuard,
                 socket_->close(ec);
             }
         }
-        if (ec == as::error::eof ||
-            ec == as::error::connection_reset
+        if (   (ec == error_code(as::error::eof, as::error::misc_category))
+            || (ec == error_code(as::error::connection_reset, as::error::system_category))
```
Comment (redboltz) [inline include/mqtt/endpoint.hpp]: What does this change mean?

File: include/mqtt/endpoint.hpp [hunk 3]
```diff
@@ -12030,7 +12028,7 @@ class endpoint : public std::enable_shared_from_this<endpoint<Mutex, LockGuard,
             () mutable {
                 if (!connected_) {
                     // offline async publish is successfully finished, because there's nothing to do.
-                    if (func) func(boost::system::errc::make_error_code(boost::system::errc::success));
+                    if (func) func({});
```
Comment (redboltz) [inline include/mqtt/endpoint.hpp]: The same as the comment above. Explicit error code is easier to understand.

File: include/mqtt/ws_endpoint.hpp [hunk 4]
```diff
@@ -39,7 +40,7 @@ class ws_endpoint {
             ws_.read(buffer, ec);
         } while (!ec);
         if (ec != boost::beast::websocket::error::closed) return;
-        ec = boost::system::errc::make_error_code(boost::system::errc::success);
+        ec = {};
```
Comment (redboltz) [inline include/mqtt/ws_endpoint.hpp]: The same as the above comment.

File: include/mqtt/ws_endpoint.hpp [hunk 5]
```diff
@@ -101,20 +102,20 @@ class ws_endpoint {
...
         using beast_read_handler_t =
-            std::function<void(boost::system::error_code const& ec, std::shared_ptr<void>)>;
+            std::function<void(error_code ec, std::shared_ptr<void>)>;
 
         std::shared_ptr<beast_read_handler_t> beast_read_handler;
         if (req_size <= buffer_.size()) {
             as::buffer_copy(buffers, buffer_.data(), req_size);
             buffer_.consume(req_size);
-            handler(boost::system::errc::make_error_code(boost::system::errc::success), req_size);
+            handler({}, req_size);
```
Comment (redboltz) [inline include/mqtt/ws_endpoint.hpp]: The same as the above comment.

File: include/mqtt/ws_endpoint.hpp [hunk 6]
```diff
@@ -141,7 +142,7 @@ class ws_endpoint {
                     }
                     as::buffer_copy(buffers, buffer_.data(), req_size);
                     buffer_.consume(req_size);
-                    std::forward<ReadHandler>(handler)(boost::system::errc::make_error_code(boost::system::errc::success), req_size);
+                    std::forward<ReadHandler>(handler)({}, req_size);
```
Comment (redboltz) [inline include/mqtt/ws_endpoint.hpp]: The same as the above comment.

File: test/connect.cpp [hunk 7]
```diff
@@ -230,15 +230,15 @@ BOOST_AUTO_TEST_CASE( keep_alive_and_send_control_packet ) {
...
                             MQTT_CHK("2sec");
                             BOOST_CHECK(!ec);
                             c->publish("topic1", "timer_reset", MQTT_NS::qos::at_most_once);
                             tim.expires_from_now(boost::posix_time::seconds(4));
                             tim.async_wait(
-                                [&chk](boost::system::error_code const& ec) {
+                                [&chk](MQTT_NS::error_code ec) {
                                     MQTT_CHK("4sec_cancelled");
-                                    BOOST_TEST(ec == boost::asio::error::operation_aborted );
+                                    BOOST_TEST(ec == MQTT_NS::error_code(boost::asio::error::make_error_code(boost::asio::error::operation_aborted)));
```
Comment (redboltz) [inline test/connect.cpp]: What does this update mean?

File: include/mqtt/client.hpp (hunk 1 above)
Comment (jonesmz) [reply include/mqtt/client.hpp]: Reverted

File: include/mqtt/endpoint.hpp (hunk 3 above)
Comment (jonesmz) [reply include/mqtt/endpoint.hpp]: Reverted

File: include/mqtt/ws_endpoint.hpp (hunk 5 above)
Comment (jonesmz) [reply include/mqtt/ws_endpoint.hpp]: Reverted

File: include/mqtt/ws_endpoint.hpp (hunk 6 above)
Comment (jonesmz) [reply include/mqtt/ws_endpoint.hpp]: Reverted

File: test/connect.cpp (hunk 7 above)
Comment (jonesmz) [reply test/connect.cpp]: Reverted

File: include/mqtt/ws_endpoint.hpp (hunk 4 above)
Comment (jonesmz) [reply include/mqtt/ws_endpoint.hpp]: Reverted

File: include/mqtt/endpoint.hpp (hunk 2 above)
Comment (jonesmz) [reply include/mqtt/endpoint.hpp]: Reverted

**Comment by jonesmz** [comment]
@redboltz did I address all of your concerns?

**Comment by redboltz** [comment]
Yes. Thanks.

**Pull request merged**